webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/domain/queries/NodeQueryCriteriaData.scala

```diff
@@ -519,6 +523,50 @@ final case class NodeCriterionMatcherString(extractor: CoreNodeFact => Chunk[Str
   val order: Ordering[String] = Ordering.String
 }
 
+/*
+ * We want to match IP address:
+ * - if equals or not equals, parse as CIDR or else IP address and work on that
+ * - if regex and other cases, parse as string
+ */
+object NodeCriterionMatcherIpaddress extends NodeCriterionMatcher {
+  private val extractor     = (n: CoreNodeFact) => n.ipAddresses.map(_.inet)
+  private val stringMatcher = NodeCriterionMatcherString(extractor)
+
+  override def matches(n: CoreNodeFact, comparator: CriterionComparator, value: String): IOResult[Boolean] = {
+    comparator match {
+      case Equals    =>
+        NodeCriterionMatcherIpaddress.buildMatcher(value, extractor(n), (ips, cidr) => ips.exists(ip => cidr.contains(ip)))
+      case NotEquals =>
+        NodeCriterionMatcherIpaddress.buildMatcher(value, extractor(n), (ips, cidr) => ips.forall(ip => !cidr.contains(ip)))
+      case c         =>
+        stringMatcher.matches(n, c, value)
+    }
+  }
+
+  /*
+   * From an input 'value' that can be a cidr or an ip, build a matcher on values
+   */
+  def buildMatcher(value: String, facts: Chunk[String], m: (Chunk[IpAddress], Cidr[IpAddress]) => Boolean): IOResult[Boolean] = {
+    Cidr
+      .fromString(value)
+      .orElse(IpAddress.fromString(value).map {
+        case ip: Ipv4Address => Cidr(ip, 32)
+        case ip: Ipv6Address => Cidr(ip, 128)
```

Comment:
when we have an IP and not a CIDR, we build the corresponding CIDR with a 0 free bits mask, which is the same as the IP address.